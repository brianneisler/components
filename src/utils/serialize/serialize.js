import { isNull } from '@serverless/utils'
import { createContext, generateReferenceKey, walkSerializeReferenceables } from './utils'

// NOTE BRN: This method belongs here and not in utils because the core has to have an understanding of how it serialized things and has to deal with backward compatability of older versions of serialized data. Will need a way of determining which serialization version was used. Later, this would probably be something that belongs in an SDK.
const serialize = (value, context) => {
  context = createContext(context)
  if (isNull(value)) {
    return null
  }
  return {
    entryKey: generateReferenceKey(context, value),
    referenceables: walkSerializeReferenceables(context, value)
  }
}

export default serialize
