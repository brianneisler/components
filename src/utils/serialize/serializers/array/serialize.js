import { isArray } from '@serverless/utils'
import serializeArray from '../../utils/serializeArray'

const serialize = (array, context) => {
  if (!isArray(array)) {
    throw new TypeError(`array.serialize expected an array, instead received ${array}`)
  }
  return {
    ...serializeArray(context, array),
    type: 'array'
  }
}

export default serialize
