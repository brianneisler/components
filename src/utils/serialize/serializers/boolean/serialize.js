import { isBoolean } from '@serverless/utils'

const serialize = (data) => {
  if (!isBoolean(data)) {
    throw new TypeError(`boolean.serialize expected a boolean, instead received ${data}`)
  }
  return {
    data,
    type: 'boolean'
  }
}

export default serialize
