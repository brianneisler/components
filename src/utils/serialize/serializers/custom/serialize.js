import hasType from '../../../type/hasType'
import getType from '../../utils/getType'
import serializeObject from '../../utils/serializeObject'

const serialize = (instance, context) => {
  if (!hasType(instance)) {
    throw new TypeError(
      `type.serialize expected an object with a type, instead received ${instance}`
    )
  }
  return {
    ...serializeObject(context, instance),
    type: getType(instance)
  }
}

export default serialize
