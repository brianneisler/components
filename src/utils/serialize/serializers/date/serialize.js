import { isDate } from '@serverless/utils'

const serialize = (data) => {
  if (!isDate(data)) {
    throw new TypeError(`date.serialize expected a Date object, instead received ${data}`)
  }
  return {
    data: data.toString(),
    type: 'date'
  }
}

export default serialize
