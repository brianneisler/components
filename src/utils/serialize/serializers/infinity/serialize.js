import { isInfinity } from '@serverless/utils'

const serialize = (data) => {
  if (!isInfinity(data)) {
    throw new TypeError(`infinity.serialize expected Infinity, instead received ${data}`)
  }
  return {
    negative: -Infinity === +data,
    type: 'infinity'
  }
}

export default serialize
