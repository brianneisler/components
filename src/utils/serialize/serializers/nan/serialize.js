import { isNaN } from '@serverless/utils'

const serialize = (data) => {
  if (!isNaN(data)) {
    throw new TypeError(`nan.serialize expected NaN, instead received ${data}`)
  }
  return {
    type: 'nan'
  }
}

export default serialize
