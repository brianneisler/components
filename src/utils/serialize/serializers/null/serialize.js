import { isNull } from '@serverless/utils'

const serialize = (data) => {
  if (!isNull(data)) {
    throw new TypeError(`null.serialize expected a null value, instead received ${data}`)
  }
  return {
    type: 'null'
  }
}

export default serialize
