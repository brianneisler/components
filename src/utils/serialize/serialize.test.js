import createContext from '../context/createContext'
import serialize from './serialize'

describe('#serialize()', () => {
  it('serialize a simple empty component', async () => {
    const component = {
      instanceId: 'foo'
    }
    const context = {}
    const result = serialize(component, context)
    expect(result).toEqual({
      entryKey: '@@ref-1',
      referenceables: {
        '@@ref-1': {
          props: {
            instanceId: {
              data: 'foo',
              type: 'string'
            }
          },
          symbols: {},
          type: 'object'
        }
      }
    })
  })

  it('serialize a simple symbol component', async () => {
    const symFoo = Symbol('foo')
    const component = {
      instanceId: 'foo',
      [symFoo]: {
        foo: 'bar'
      }
    }
    const context = {
      log: () => {},
      symbolMap: {
        [symFoo]: '@@foo'
      }
    }
    const result = serialize(component, context)
    expect(result).toEqual({
      entryKey: '@@ref-1',
      referenceables: {
        '@@ref-1': {
          props: {
            instanceId: {
              data: 'foo',
              type: 'string'
            }
          },
          symbols: {
            '@@foo': {
              '@@ref': '@@ref-2'
            }
          },
          type: 'object'
        },
        '@@ref-2': {
          props: {
            foo: {
              data: 'bar',
              type: 'string'
            }
          },
          symbols: {},
          type: 'object'
        }
      }
    })
  })

  it('serialize a simple component', async () => {
    const context = await createContext({}, { app: { id: 'test' } })
    const Component = await context.import('Component')
    const component = context.construct(Component, {})
    component.instanceId = 'test'
    const result = serialize(component, context)
    expect(result).toMatchSnapshot()
  })
})
