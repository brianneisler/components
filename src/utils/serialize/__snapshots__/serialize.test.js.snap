// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#serialize() serialize a simple component 1`] = `
Object {
  "entryKey": "@@ref-1",
  "referenceables": Object {
    "@@ref-1": Object {
      "props": Object {
        "components": Object {
          "@@ref": "@@ref-2",
        },
        "extends": Object {
          "data": "Object",
          "type": "string",
        },
        "inputTypes": Object {
          "@@ref": "@@ref-3",
        },
        "inputs": Object {
          "@@ref": "@@ref-4",
        },
        "instanceId": Object {
          "data": "test",
          "type": "string",
        },
        "main": Object {
          "data": "./dist/index.js",
          "type": "string",
        },
        "name": Object {
          "data": "Component",
          "type": "string",
        },
        "version": Object {
          "data": "0.3.0",
          "type": "string",
        },
      },
      "symbols": Object {},
      "type": "Component",
    },
    "@@ref-2": Object {
      "props": Object {},
      "symbols": Object {},
      "type": "object",
    },
    "@@ref-3": Object {
      "props": Object {
        "components": Object {
          "@@ref": "@@ref-5",
        },
      },
      "symbols": Object {},
      "type": "object",
    },
    "@@ref-4": Object {
      "props": Object {},
      "symbols": Object {},
      "type": "object",
    },
    "@@ref-5": Object {
      "props": Object {
        "default": Object {
          "@@ref": "@@ref-6",
        },
        "type": Object {
          "data": "Object",
          "type": "string",
        },
      },
      "symbols": Object {},
      "type": "object",
    },
    "@@ref-6": Object {
      "props": Object {},
      "symbols": Object {},
      "type": "object",
    },
  },
}
`;
